### ISSUE TITLE
Feature request: make index file parameterisable for VariantFile (and others?)

### ISSUE THREAD

**Issue by jeromekelleher** [pysam-developers/pysam#87]
Currently, when we open a VariantFile from a HTTP or FTP URL, the low-level htslib code checks for the existence of the corresponding index file, and writes it to the current working directory, if it exists. This behaviour is perfectly fine for many applications, but is not suitable when many different files may be opened at the same time, which may happen to have the same index names. It is also not very good for the GA4GH server, where we want to allow HTTP URLs to be allowed as a data source (see https://github.com/ga4gh/server/pull/232 for details). We don't want our server blocking for an arbitrary amount of time while it downloads index files, and also having an apache process writing lots of essential files (possibly in parallel, overwriting and corrupting partially downloaded files, since there will be several processes) to the file system.

It would be much better if we could pre-download the index files, and then provide a base URL from which we can find the data file. This should be straightforward enough to do, but we'd need to have the option of specifying the index file to VariantFile when we open the URL (and others; presumably, we'll want to do the same thing for SAM files too).

Is this feasible? I don't mind doing the legwork if you give me some pointers on where to start.

**Comment by AndreasHeger** [comment]
Thanks, that makes sense. I will see what is involved.

Best wishes

**Comment by bioinformed** [comment]
I think this will be easy.  I'm looking into it too.

**Comment by AndreasHeger** [comment]
Thanks!

**Comment by AndreasHeger** [comment]
Hi @jeromekelleher , @bioinformed  - I have not followed this up - anything I can do to help?

Best wishes,
Andreas

**Comment by jeromekelleher** [comment]
@AndreasHeger, @bioinformed, is there anything I can do to help here? If some functionality is required in htslib, I can open an issue there.

**Comment by bioinformed** [comment]
There does need to be a small change to htslib.  We need a version of
hts_idx_load that takes an explicit index filename.  The current version
infers the index name from the data file name.  (The tabix API doesn't have
this extra logic.)

This has been on my TODO list, but I've been crushed with travel and work.

Thanks,
-Kevin

**Comment by AndreasHeger** [comment]
Thanks Kevin. I agree it would be best if htslib would provide that functionality. I find it a bit hard to make sense of the discussion, there seems to be a C++ patch?

@jeromekelleher , how critical is this - can we wait until there is an htslib solution?

I would also be happy to implement a hts_idx_load drop-in for pysam that will take an optional filename (and if none given, just call htslib hts_idx_load).

Best, Andreas

**Comment by jeromekelleher** [comment]
Thanks @AndreasHeger and @bioinformed. There isn't a huge time pressure on this, so I think it would be better if we could get htslib to provide the required functionality. Anyway, you guys are doing enough! 

Is there an issue open on this for htslib, or shall I open it?

**Comment by AndreasHeger** [comment]
Dear both,
I had a quick go at this - at least for bam/cram files. This section works without changes to htslib.

Please see branch AH-IndexFilename. It is now possible to use a local index for a remote file:

```python

    def testRemoteExplicitIndexBAM(self):
        samfile = pysam.AlignmentFile(
            "http://genserv.anat.ox.ac.uk/downloads/pysam/test/noindex.bam",
            "rb",
            filepath_index=os.path.join(DATADIR, 'ex1.bam.bai'))

        samfile.fetch("chr1")
```

@jeromekelleher, would this cover your use case?
Best, Andreas

**Comment by jeromekelleher** [comment]
Thanks @AndreasHeger, this looks great. This API would cover our use-case perfectly. However, I wonder if there's much point in pushing these changes through, since the modifications to htslib are done and just waiting to be cleaned up and merged --- is this correct @jmarshall?

**Comment by AndreasHeger** [comment]
Hi @jeromekelleher, the above is possible without any upstream changes in htslib. The call to open an index file is a separate method and accepts a filename. Giving a different filename for the index seems to work without problems (there are some constraints on the suffix (.bai, etc).
Best, Andreas

**Comment by jmarshall** [comment]
> Giving a different filename for the index seems to work without problems (there are some constraints on the suffix (.bai, etc).

You are giving `sam_index_load` a `cfilename` that you know doesn't exist, which happens to mostly work at the moment, but is not a supported way of using `sam_index_load`.  For BAM, you seem to be giving `sam_index_load` the filename of the .bai file, which wouldn't work at all.

It's bad enough that we all have all this code to search for indices at various filenames alongside a BAM file; please let's not add code to reconstruct a BAM/CRAM filename from an index filename in the other direction as well!

Apologies for the delay… I have redesigned the explicit-index-filename htslib API half a dozen times in the last week!  However finally HTSlib additions that will add

```c
/// Load a specific index file
/** @param fp     File handle of the data file whose index is being opened
    @param fn     BAM/CRAM/etc data file filename
    @param fnidx  Index filename, or NULL to search alongside @a fn
    @return  The index, or NULL if an error occurred.
*/  
hts_idx_t *sam_index_load2(htsFile *fp, const char *fn, const char *fnidx);
```

are coming this afternoon.

**Comment by AndreasHeger** [comment]
Thanks. An explicit index filename would indeed be better, but it seems that the current sam_index_load and hts_idx_load seem to work (tested for BAM/CRAM and also remote BAM).
Best,
Andreas

**Comment by jmarshall** [comment]
So it does, for now — by pure luck.  HTSlib maintainers recommend not taking advantage of this.

**Comment by AndreasHeger** [comment]
Thanks, I will wait. Best, Andreas

**Comment by bioinformed** [comment]
I'll create a branch with the latest htslib and will update the bindings to support all of the new and changed APIs.  Once that is merged, we can properly address this issue.

**Comment by AndreasHeger** [comment]
The AH-IndexFilename branch has now been merged into pysam-0.8.5-dev branch and will soon be released.

**Comment by jeromekelleher** [comment]
I've tested out this functionality on BAM files @AndreasHeger (see https://github.com/ga4gh/server/pull/880 ), and it works great; thanks! 

I can't seem to get the equivalent idea to work for VCFs. Is this supported?

**Comment by bioinformed** [comment]
I'll implement a similar system for the new BCF/VCF API.

**Comment by jeromekelleher** [comment]
Thanks @bioinformed, that would be great.

**Comment by AndreasHeger** [comment]
Closing for now. Implemented
